src/main/kotlin/org/usecase/dao/PrincipalDao.kt

```diff
@@ -14,23 +15,23 @@ class PrincipalDao(val con: AbstractConnector) {
     fun getOne(idPrincipalPk: Long): Principal? {
         // TODO: review generated method
         val query = Query()
-                .selectAll()
+                .selectFields(Principal.selectFields())
                 .from("principal")
                 .whereEq("idPrincipalPk", idPrincipalPk)
 
         return con.getOne(query) {
-            Principal(it)
+            Principal(ResultBuilder(Principal.selectFields(), it, "principal"))
```

Comment:
First, I have a point that I didn't get. Why does ResultBuilder need to define the selected columns? You have selected the columns during the query. Isn't that enough already? In my view, it is a duplicated definition that can be avoided. 

Second, I have a suggestion to improve the code organization. Tell me what you think.

_Creating a package called builder which has builder models_

Example:
**PrincipalBuilder.kt**
```kotlin
package org.usecase.model.builder

import br.com.simpli.sql.ResultBuilder
import org.usecase.model.resource.Principal
import java.sql.ResultSet

class PrincipalBuilder(val rs: ResultSet, val alias: String = "principal") : ResultBuilder(selectFields(alias), rs, alias) {
    fun build(): Principal {
        return Principal().apply {
            idPrincipalPk = getLong("idPrincipalPk")
            textoObrigatorio = getString("textoObrigatorio")
            textoFacultativo = getString("textoFacultativo")
            decimalObrigatorio = getDouble("decimalObrigatorio")
            decimalFacultativo = getDoubleOrNull("decimalFacultativo")
            inteiroObrigatorio = getLong("inteiroObrigatorio")
            inteiroFacultativo = getLongOrNull("inteiroFacultativo")
            booleanoObrigatorio = getBoolean("booleanoObrigatorio")
            booleanoFacultativo = getBooleanOrNull("booleanoFacultativo")
            dataObrigatoria = getTimestamp("dataObrigatoria")
            dataFacultativa = getTimestamp("dataFacultativa")
            datahoraObrigatoria = getTimestamp("datahoraObrigatoria")
            datahoraFacultativa = getTimestamp("datahoraFacultativa")
            ativo = getBoolean("ativo")
            email = getString("email")
            senha = getString("senha")
            urlImagem = getString("urlImagem")
            url = getString("url")
            idGrupoDoPrincipalFk = getLong("idGrupoDoPrincipalFk")
            idGrupoDoPrincipalFacultativoFk = getLongOrNull("idGrupoDoPrincipalFacultativoFk")
            unico = getString("unico")
            dataCriacao = getTimestamp("dataCriacao")
            dataAlteracao = getTimestamp("dataAlteracao")
            nome = getString("nome")
            titulo = getString("titulo")
            cpf = getString("cpf")
            cnpj = getString("cnpj")
            rg = getString("rg")
            celular = getString("celular")
            textoGrande = getString("textoGrande")
            snakeCase = getString("snake_case")
            preco = getDoubleOrNull("preco")
        }
    }

    companion object {
        fun selectFields(alias: String = "principal") = arrayOf(
                "$alias.idPrincipalPk",
                "$alias.textoObrigatorio",
                "$alias.textoFacultativo",
                "$alias.decimalObrigatorio",
                "$alias.decimalFacultativo",
                "$alias.inteiroObrigatorio",
                "$alias.inteiroFacultativo",
                "$alias.booleanoObrigatorio",
                "$alias.booleanoFacultativo",
                "$alias.dataObrigatoria",
                "$alias.dataFacultativa",
                "$alias.datahoraObrigatoria",
                "$alias.datahoraFacultativa",
                "$alias.ativo",
                "$alias.email",
                "$alias.urlImagem",
                "$alias.url",
                "$alias.idGrupoDoPrincipalFk",
                "$alias.idGrupoDoPrincipalFacultativoFk",
                "$alias.unico",
                "$alias.dataCriacao",
                "$alias.dataAlteracao",
                "$alias.nome",
                "$alias.titulo",
                "$alias.cpf",
                "$alias.cnpj",
                "$alias.rg",
                "$alias.celular",
                "$alias.textoGrande",
                "$alias.snake_case",
                "$alias.preco"
        )

        fun orderMap(alias: String = "principal") = mapOf(
                "idPrincipalPk" to "$alias.idPrincipalPk",
                "textoObrigatorio" to "$alias.textoObrigatorio",
                "textoFacultativo" to "$alias.textoFacultativo",
                "decimalObrigatorio" to "$alias.decimalObrigatorio",
                "decimalFacultativo" to "$alias.decimalFacultativo",
                "inteiroObrigatorio" to "$alias.inteiroObrigatorio",
                "inteiroFacultativo" to "$alias.inteiroFacultativo",
                "booleanoObrigatorio" to "$alias.booleanoObrigatorio",
                "booleanoFacultativo" to "$alias.booleanoFacultativo",
                "dataObrigatoria" to "$alias.dataObrigatoria",
                "dataFacultativa" to "$alias.dataFacultativa",
                "datahoraObrigatoria" to "$alias.datahoraObrigatoria",
                "datahoraFacultativa" to "$alias.datahoraFacultativa",
                "ativo" to "$alias.ativo",
                "email" to "$alias.email",
                "urlImagem" to "$alias.urlImagem",
                "url" to "$alias.url",
                "idGrupoDoPrincipalFk" to "$alias.idGrupoDoPrincipalFk",
                "idGrupoDoPrincipalFacultativoFk" to "$alias.idGrupoDoPrincipalFacultativoFk",
                "unico" to "$alias.unico",
                "dataCriacao" to "$alias.dataCriacao",
                "dataAlteracao" to "$alias.dataAlteracao",
                "nome" to "$alias.nome",
                "titulo" to "$alias.titulo",
                "cpf" to "$alias.cpf",
                "cnpj" to "$alias.cnpj",
                "rg" to "$alias.rg",
                "celular" to "$alias.celular",
                "textoGrande" to "$alias.textoGrande",
                "snakeCase" to "$alias.snake_case",
                "preco" to "$alias.preco"
        )
    }
}
```

**Result in DAO:**
```kotlin
    fun getOne(idPrincipalPk: Long): Principal? {
        // TODO: review generated method
        val query = Query()
                .selectFields(PrincipalBuilder.selectFields())
                .from("principal")
                .whereEq("idPrincipalPk", idPrincipalPk)

        return con.getOne(query) {
            PrincipalBuilder(it).build()
        }
    }
```

Reply:
We already open a discussion on discord, I think you missed that.
Anyway, I will talk about your suggestions on discord, but I prefer the way it is.